File: dp/3098-find-the-sum-of-subsequence-powers.py
from typing import List
import unittest

# https://leetcode.com/problems/find-the-sum-of-subsequence-powers/

# python3 -m unittest dp/3098-find-the-sum-of-subsequence-powers.py


class Solution(unittest.TestCase):
    # Approach: Top-Down Dynamic Programming
    # Time: O(nnn k d), n=len(nums), d=diff
    # Space: O(nkd)
    def sumOfPowers(self, nums: List[int], k: int) -> int:
        nums.sort()
        memo = {}
        MOD = 10**9 + 7
        n = len(nums)
        def dp(idx: int, k: int, diff: int) -> int:
            if k == 0:
                return diff
            key = (idx, k, diff)
            if key in memo:
                return memo[key]
            total = 0
            for nxt in range(idx+1, n):
                total = (total + dp(nxt, k-1, min(diff, nums[nxt] - nums[idx]))) % MOD
            memo[key] = total
            return total
        return sum(dp(i, k-1, (1 << 32) - 1) for i in range(n)) % MOD

    def test(self):
        for nums, k, expected in [
            ([1,2,3,4], 3, 4),
            ([2,2], 2, 0),
            ([4,3,-1], 2, 10),
            # ([-24, -921, 119, -291, -65, -628, 372, 274, 962, -592, -10, 67, -977, 85, -294, 349, -119, -846, -959, -79, -877, 833, 857, 44, 826, -295, -855, 554, -999, 759, -653, -423, -599, -928], 19, 990202285),
        ]:
            output = self.sumOfPowers(nums, k)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: dp/3098-find-the-sum-of-subsequence-powers.go
package dp

import "sort"

// https://leetcode.com/problems/find-the-sum-of-subsequence-powers/

// Approach #2: Top-Down Dynamic Programming
// Time: O(nn k d), n=len(nums), d=diff
// Space: O(nn k d)
func sumOfPowers(nums []int, k int) int {
	sort.Ints(nums)
	const mod = int(1e9) + 7
	n := len(nums)
	memo := make(map[[3]int]*int)
	var dp func(idx int, k int, diff int) int
	dp = func(idx, k, diff int) int {
		if k == 0 {
			return diff
		}
		key := [3]int{idx, k, diff}
		if memo[key] != nil {
			return *memo[key]
		}
		var total int
		for nxt := idx + 1; nxt < n; nxt++ {
			total = (total + dp(nxt, k-1, min(diff, nums[nxt]-nums[idx]))) % mod
		}
		memo[key] = &total
		return total
	}
	inf := (1 << 32) - 1
	var total int
	for i := range n {
		total = (total + dp(i, k-1, inf)) % mod
	}
	return total
}

// // Approach #1: Top-Down Dynamic Programming (slower)
// // Time: O(nn k d), n=len(nums), d=diff
// // Space: O(nn k d)
// func sumOfPowers(nums []int, k int) int {
// 	sort.Ints(nums)
// 	const mod = int(1e9) + 7
// 	memo := make(map[[4]int]*int)
// 	var dp func(prev int, curr int, k int, diff int) int
// 	dp = func(prev, curr, k, diff int) int {
// 		if k == 0 {
// 			return diff
// 		}
// 		if curr == len(nums) {
// 			return 0
// 		}
// 		key := [4]int{prev, curr, k, diff}
// 		if memo[key] != nil {
// 			return *memo[key]
// 		}
// 		// skip
// 		total := dp(prev, curr+1, k, diff)
// 		// include
// 		if prev >= 0 {
// 			diff = min(diff, nums[curr]-nums[prev])
// 		}
// 		total = (total + dp(curr, curr+1, k-1, diff)) % mod
// 		memo[key] = &total
// 		return total
// 	}
// 	return dp(-1, 0, k, (1<<32)-1)
// }
